dask/dataframe/io/hdf.py

```diff
@@ -268,142 +271,43 @@ def to_hdf(
-                    stops.append(stop)
-                if sorted_index:
-                    division = [
-                        storer.read_column("index", start=start, stop=start + 1)[0]
-                        for start in range(0, storer.nrows, chunksize)
-                    ]
-                    division_end = storer.read_column(
-                        "index", start=storer.nrows - 1, stop=storer.nrows
-                    )[0]
-
-                    division.append(division_end)
-                    divisions.append(division)
-                else:
-                    divisions.append(None)
-
-        return keys, stops, divisions
-
-    def one_path_one_key(path, key, start, stop, columns, chunksize, division, lock):
-        """
-        Get the data frame corresponding to one path and one key (which should
-        not contain any wildcards).
+    def project_columns(self, columns):
+        """Return a new HDFFunctionWrapper object with
+        a sub-column projection.
         """
-        empty = pd.read_hdf(path, key, mode=mode, stop=0)
-        if columns is not None:
-            empty = empty[columns]
-
-        token = tokenize(
-            (path, os.path.getmtime(path), key, start, stop, empty, chunksize, division)
-        )
-        name = "read-hdf-" + token
-        if empty.ndim == 1:
-            base = {"name": empty.name, "mode": mode}
-        else:
-            base = {"columns": empty.columns, "mode": mode}
-
-        if start >= stop:
-            raise ValueError(
-                "Start row number ({}) is above or equal to stop "
-                "row number ({})".format(start, stop)
-            )
-
-        def update(s):
-            new = base.copy()
-            new.update({"start": s, "stop": s + chunksize})
-            return new
-
-        dsk = dict(
-            ((name, i), (_pd_read_hdf, path, key, lock, update(s)))
-            for i, s in enumerate(range(start, stop, chunksize))
-        )
-
-        if division:
-            divisions = division
-        else:
-            divisions = [None] * (len(dsk) + 1)
-
-        return new_dd_object(dsk, name, empty, divisions)
-
-    keys, stops, divisions = get_keys_stops_divisions(
-        path, key, stop, sorted_index, chunksize
-    )
-    if (start != 0 or stop is not None) and len(keys) > 1:
-        raise NotImplementedError(read_hdf_error_msg)
-    from ..multi import concat
-
-    return concat(
-        [
-            one_path_one_key(path, k, start, s, columns, chunksize, d, lock)
-            for k, s, d in zip(keys, stops, divisions)
-        ]
-    )
-
-
-def _pd_read_hdf(path, key, lock, kwargs):
-    """ Read from hdf5 file with a lock """
-    if lock:
-        lock.acquire()
-    try:
-        result = pd.read_hdf(path, key, **kwargs)
-    finally:
-        if lock:
-            lock.release()
-    return result
+        if columns == self.columns:
+            return self
+        func = copy.deepcopy(self)
```

Comment:
I see this is the same pattern you've used in the other dataframe IO layers, though I missed it in those. Is there a reason you have used `deepcopy` here rather than the constructor? It feels more future-proof to me to just call `HDFFunctionWrapper`, which might some day do other logic on the column listing not expressible by copying.

Reply:
Thanks for pointing this out - There was once a (probably silly) reason for doing this.  However, it seems that the complication has been resolved.   I just removed the deep copy throughout.